database/src/tests/mod.rs

```diff
@@ -0,0 +1,103 @@
+    db_name: String,
+    original_db_url: String,
+    // Pre-cached client to avoid creating unnecessary connections in tests
+    client: Pool,
+}
+
+impl TestContext {
+    async fn new(db_url: &str) -> Self {
+        let config: Config = db_url.parse().expect("Cannot parse connection string");
+
+        // Create a new database that will be used for this specific test
+        let client = make_client(&db_url)
+            .await
+            .expect("Cannot connect to database");
+        let db_name = format!("db{}", uuid::Uuid::new_v4().to_string().replace("-", ""));
+        client
+            .execute(&format!("CREATE DATABASE {db_name}"), &[])
+            .await
+            .expect("Cannot create database");
+        drop(client);
+
+        let host = match &config
+            .get_hosts()
+            .first()
+            .expect("connection string must contain at least one host")
+        {
+            Host::Tcp(host) => host,
+
+            // This variant only exists on Unix targets, so the arm itself is
+            // cfg-gated to keep non-unix builds happy.
+            #[cfg(unix)]
+            Host::Unix(_) => panic!("Unix sockets in Postgres connection string are not supported"),
+
+            // On non-unix targets the enum has no other variants.
+            #[cfg(not(unix))]
+            _ => unreachable!("non-TCP hosts cannot appear on this platform"),
+        };
+
+        // We need to connect to the database against, because Postgres doesn't allow
+        // changing the active database mid-connection.
+        // There does not seem to be a way to turn the config back into a connection
+        // string, so construct it manually.
+        let test_db_url = format!(
+            "postgresql://{}:{}@{}:{}/{}",
+            config.get_user().unwrap(),
+            String::from_utf8(config.get_password().unwrap().to_vec()).unwrap(),
+            host,
+            &config.get_ports()[0],
+            db_name
+        );
+        let pool = Pool::open(test_db_url.as_str());
+
+        Self {
+            db_name,
+            original_db_url: db_url.to_string(),
+            client: pool,
+        }
+    }
+
+    pub(crate) fn db_client(&self) -> &Pool {
+        &self.client
+    }
+
+    async fn finish(self) {
+        // Cleanup the test database
+        // First, we need to stop using the database
+        drop(self.client);
+
+        // Then we need to connect to the default database and drop our test DB
+        let client = make_client(&self.original_db_url)
+            .await
+            .expect("Cannot connect to database");
+        client
+            .execute(&format!("DROP DATABASE {}", self.db_name), &[])
+            .await
+            .unwrap();
+    }
+}
+
+pub(crate) async fn run_db_test<F, Fut>(f: F)
+where
+    F: FnOnce(TestContext) -> Fut,
+    Fut: Future<Output = anyhow::Result<TestContext>>,
+{
+    if let Ok(db_url) = std::env::var("TEST_DB_URL") {
```

Comment:
Let's just crash here if `TEST_DB_URL` is unset. In triagebot there were some existing tests that I didn't want to break, but I don't think that many people run `cargo test` on the rustc-perf, and it's safer to just panic if the DB is missing, to avoid a situation where tests are being skipped unexpectedly.

Reply:
I have `cfg`'d this as the pipeline failed when running windows as it does not support a docker container. Thus, how I've set things up is to skip the database tests on Windows.